File: frontend/src/types/course.ts
/**
 * TypeScript types for Course domain.
 *
 * Defines types for courses, modules, content items, and related entities.
 */

/**
 * Course visibility options.
 */
export enum CourseVisibility {
  PUBLIC = 'public',
  PRIVATE = 'private',
  DRAFT = 'draft',
}

/**
 * Content type options.
 */
export enum ContentType {
  VIDEO = 'video',
  DOCUMENT = 'document',
  TEXT = 'text',
  QUIZ = 'quiz',
  ASSIGNMENT = 'assignment',
}

/**
 * Course category (common categories).
 */
export type Category =
  | 'programming'
  | 'design'
  | 'business'
  | 'marketing'
  | 'data-science'
  | 'personal-development'
  | 'other'

/**
 * Base course interface.
 */
export interface Course {
  id: string
  instructor_id: string
  title: string
  description: string
  category: string
  visibility: CourseVisibility
  enrollment_capacity: number | null
  thumbnail_url: string | null
  is_featured: boolean
  created_at: string
  updated_at: string
  is_published: boolean
  is_public: boolean
}

/**
 * Course with instructor details.
 */
export interface CourseWithInstructor extends Course {
  instructor_name: string
  instructor_email: string
}

/**
 * Course creation request.
 */
export interface CourseCreate {
  title: string
  description: string
  category: string
  visibility?: CourseVisibility
  enrollment_capacity?: number | null
  thumbnail_url?: string | null
  is_featured?: boolean
}

/**
 * Course update request.
 */
export interface CourseUpdate {
  title?: string
  description?: string
  category?: string
  visibility?: CourseVisibility
  enrollment_capacity?: number | null
  thumbnail_url?: string | null
  is_featured?: boolean
}

/**
 * Paginated course list response.
 */
export interface CourseListResponse {
  courses: Course[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

/**
 * Module for organizing course content.
 */
export interface Module {
  id: string
  course_id: string
  title: string
  description: string | null
  order_index: number
  created_at: string
  updated_at: string
}

/**
 * Module creation request.
 */
export interface ModuleCreate {
  course_id: string
  title: string
  description?: string | null
  order_index: number
}

/**
 * Module update request.
 */
export interface ModuleUpdate {
  title?: string
  description?: string | null
  order_index?: number
}

/**
 * Content item (video, document, text, etc.).
 */
export interface ContentItem {
  id: string
  course_id: string
  module_id: string | null
  title: string
  description: string | null
  content_type: ContentType
  file_url: string | null
  file_size: number | null
  mime_type: string | null
  text_content: string | null
  duration_seconds: number | null
  order_index: number
  is_required: boolean
  created_at: string
  updated_at: string
  is_file_based: boolean
  is_text_based: boolean
}

/**
 * Content item creation request.
 */
export interface ContentItemCreate {
  course_id: string
  module_id?: string | null
  title: string
  description?: string | null
  content_type: ContentType
  text_content?: string | null
  duration_seconds?: number | null
  order_index: number
  is_required?: boolean
}

/**
 * Content item update request.
 */
export interface ContentItemUpdate {
  title?: string
  description?: string | null
  text_content?: string | null
  duration_seconds?: number | null
  order_index?: number
  is_required?: boolean
}

/**
 * Content upload response.
 */
export interface ContentItemUploadResponse {
  content_item: ContentItem
  upload_url: string | null
}

/**
 * Form data for content upload.
 */
export interface ContentUploadFormData {
  file: File
  course_id: string
  title: string
  content_type: ContentType
  order_index: number
  description?: string
  module_id?: string
  text_content?: string
  duration_seconds?: number
  is_required?: boolean
}

File: backend/src/schemas/course.py
"""
Pydantic schemas for Course API.

Request and response models for course operations.
"""
from datetime import datetime
from pydantic import BaseModel, Field, field_validator
from typing import Optional

from src.models.course import CourseVisibility


class CourseBase(BaseModel):
    """Base course schema with common fields."""
    title: str = Field(..., min_length=3, max_length=200, description="Course title")
    description: str = Field(..., min_length=10, description="Course description")
    category: str = Field(..., min_length=2, max_length=100, description="Course category")
    visibility: CourseVisibility = Field(default=CourseVisibility.DRAFT, description="Course visibility")
    enrollment_capacity: Optional[int] = Field(None, ge=1, description="Maximum enrollment (null = unlimited)")
    thumbnail_url: Optional[str] = Field(None, max_length=500, description="Course thumbnail URL")
    is_featured: bool = Field(default=False, description="Whether course is featured")


class CourseCreate(CourseBase):
    """Schema for creating a new course."""

    @field_validator('title')
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        """Validate title is not just whitespace."""
        if not v.strip():
            raise ValueError("Title cannot be empty or whitespace")
        return v.strip()

    @field_validator('description')
    @classmethod
    def description_not_empty(cls, v: str) -> str:
        """Validate description is not just whitespace."""
        if not v.strip():
            raise ValueError("Description cannot be empty or whitespace")
        return v.strip()


class CourseUpdate(BaseModel):
    """Schema for updating an existing course."""
    title: Optional[str] = Field(None, min_length=3, max_length=200)
    description: Optional[str] = Field(None, min_length=10)
    category: Optional[str] = Field(None, min_length=2, max_length=100)
    visibility: Optional[CourseVisibility] = None
    enrollment_capacity: Optional[int] = Field(None, ge=1)
    thumbnail_url: Optional[str] = Field(None, max_length=500)
    is_featured: Optional[bool] = None


class CourseResponse(CourseBase):
    """Schema for course response."""
    id: str
    instructor_id: str
    created_at: datetime
    updated_at: datetime

    # Computed fields
    is_published: bool = Field(description="Whether course is published")
    is_public: bool = Field(description="Whether course is publicly accessible")

    model_config = {"from_attributes": True}


class CourseWithInstructor(CourseResponse):
    """Schema for course response with instructor details."""
    instructor_name: str = Field(description="Instructor full name")
    instructor_email: str = Field(description="Instructor email")

    model_config = {"from_attributes": True}


class CourseListResponse(BaseModel):
    """Schema for paginated course list response."""
    courses: list[CourseResponse]
    total: int = Field(description="Total number of courses")
    page: int = Field(description="Current page number")
    per_page: int = Field(description="Items per page")
    total_pages: int = Field(description="Total number of pages")

    model_config = {"from_attributes": True}
